pulser-core/pulser/math/abstract_array.py

```diff
@@ -0,0 +1,301 @@
+# Copyright 2024 Pulser Development Team
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Defines the AbstractArray class."""
+from __future__ import annotations
+
+import functools
+import importlib.util
+import operator
+from typing import Any, Generator, Union, cast
+
+import numpy as np
+from numpy.typing import ArrayLike, DTypeLike
+
+from pulser.json.utils import obj_to_dict
+
+try:
+    import torch
+except ImportError:  # pragma: no cover
+    pass
+
+
+class AbstractArray:
+    """An abstract array containing an array or tensor.
+
+    Args:
+        array: The array to store.
+    """
+
+    def __init__(
+        self,
+        array: AbstractArrayLike,
+        dtype: DTypeLike = None,
+        force_array: bool = False,
+    ):
+        """Initializes a new AbstractArray."""
+        self._array: np.ndarray | torch.Tensor
+        if isinstance(array, AbstractArray):
+            self._array = array._array
+        elif self.has_torch() and isinstance(array, torch.Tensor):
+            self._array = torch.as_tensor(
+                array,
+                dtype=dtype,  # type: ignore[arg-type]
+            )
+        else:
+            self._array = np.asarray(array, dtype=dtype)
+
+        if force_array and self._array.ndim == 0:
+            self._array = self._array[None]
+
+    @staticmethod
+    @functools.lru_cache
+    def has_torch() -> bool:
+        """Checks whether torch is installed."""
+        return importlib.util.find_spec("torch") is not None
```

Comment:
`find_spec()` does not import the module, so we'd still need the `import torch`. We could make it conditional on `HAS_TORCH` but I don't see how that would be an improvement tbh.
I think this approach is more robust because the value of `has_torch()` is protected, unlike with `HAS_TORCH` being a constant.
Finally, I'm afraid the typing issues would not resolve because all typing checks are done statically ie the code is not executed. Therefore, the static type checkers won't know what `HAS_TORCH` is and they don't evaluate `if` statements. I suspect defining the types as you proposed will even throw an error.